lib/app/features/chat/providers/private_direct_chat_provider.c.dart

```diff
@@ -0,0 +1,370 @@
+          log('Compressed media file: ${compressedMediaFile.path}, ${compressedMediaFile.mimeType} $compressedSize');
+
+          return compressedMediaFile;
+        },
+      ).toList(),
+    );
+
+    return compressedMediaFiles;
+  }
+
+  Future<List<(MediaFile, String, String)>> _encryptMediaFiles(
+    List<MediaFile> compressedMediaFiles,
+  ) async {
+    final encryptedMediaFilesWithSecretBox = await Future.wait(
+      compressedMediaFiles.map(
+        (compressedMediaFile) => Isolate.run<(MediaFile, String, String)>(() async {
+          final secretKey = await AesGcm.with256bits().newSecretKey();
+          final secretKeyBytes = await secretKey.extractBytes();
+          final secretKeyString = base64Encode(secretKeyBytes);
+
+          final compressedMediaFileBytes = await File(compressedMediaFile.path).readAsBytes();
+
+          final secretBox = await AesGcm.with256bits().encrypt(
+            compressedMediaFileBytes,
+            secretKey: secretKey,
+          );
+
+          final nonceBytes = secretBox.nonce;
+          final nonceString = base64Encode(nonceBytes);
+
+          final compressedEncryptedFile = File('${compressedMediaFile.path}.enc');
+          // Rewrite compressed fieles with encrypted data
+          await compressedEncryptedFile.writeAsBytes(secretBox.cipherText);
+
+          final compressedEncryptedMediaFile = MediaFile(
+            path: compressedEncryptedFile.path,
+            width: compressedMediaFile.width,
+            height: compressedMediaFile.height,
+            mimeType: compressedMediaFile.mimeType,
+          );
+
+          log('Encrypted media file ${compressedEncryptedMediaFile.mimeType} ${compressedEncryptedFile.lengthSync()}');
+
+          return (compressedEncryptedMediaFile, nonceString, secretKeyString);
+        }),
+      ),
+    );
+
+    return encryptedMediaFilesWithSecretBox;
+  }
+
+  List<List<String>> _generateImetaTags(
+    List<(UploadResult, String, String)> uploadResults,
+  ) {
+    final expiration = EntityExpiration(
+      value: DateTime.now().add(
+        Duration(
+          hours: ref.read(envProvider.notifier).get<int>(EnvVariable.STORY_EXPIRATION_HOURS),
+        ),
+      ),
+    );
+
+    return uploadResults.map((uploadResult) {
+      final fileMetadata = uploadResult.$1.fileMetadata;
+
+      final nonce = uploadResult.$2;
+      final secretKey = uploadResult.$3;
+
+      return [
+        'imeta',
+        'url ${fileMetadata.url}',
+        'alt ${fileMetadata.alt}',
+        'm ${fileMetadata.mimeType}',
+        'dim ${fileMetadata.dimension}',
+        'x ${fileMetadata.fileHash}',
+        'ox ${fileMetadata.originalFileHash}',
+        'expiration ${expiration.value.millisecondsSinceEpoch ~/ 1000}',
+        'encryption-key $secretKey $nonce aes-gcm',
+      ];
+    }).toList();
+  }
+}
+
+// Example usage:
+/*
+  final privateDirectChatProvider = ref.read(privateDirectChatProviderProvider.notifier);
+
```

Comment:
Maybe move it to the beginning of the file?

Reply:
Done